Add Joblib Launcher for parallel execution (multi-process and multi-threaded) (#392)
facebookresearch/hydra: plugins/hydra_joblib_launcher/README.md

    backend: null  # allows to hard-code backend, otherwise inferred based on prefer and require
    prefer: processes  # processes or threads, soft hint to choose backend
    require: null  # null or sharedmem, sharedmem will select thread-based backend
    verbose: 0  # if greater than zero, prints progress messages
    timeout: null  # timeout limit for each task
    pre_dispatch: 2*n_jobs  #  number of batches to be pre-dispatched
    batch_size: auto  # number of atomic tasks to dispatch at once to each worker
    temp_folder: null  # folder used for memmapping large arrays for sharing memory with workers
    max_nbytes: 1M  # thresholds size of arrays that triggers automated memmapping
    mmap_mode: r  # memmapping mode for numpy arrays passed to workers
```

All arguments specified in `joblib` are passed to `Joblib.Parallel` (see [`Joblib.Parallel` documentation](https://joblib.readthedocs.io/en/latest/parallel.html) for details). `n_jobs` defaults to -1, which means that all available CPUs may be used. `prefer` defaults to `processes`, depending on the application, `threads` can be an alternative. 

> An example application using the plugin is provided in `plugins/hydra_joblib_launcher/example`. It overwrites the launcher used by Hydra.

Reviewer: Can you make the example app a link to the app dir?

```suggestion
An example application using this launcher is provided in `plugins/hydra_joblib_launcher/example`.  It changes the default launcher Hydra is using to this one.
```

Author: Done

Reviewer: You didn't delete the line this was editing.

Author: Changed `README.md` to say:

```markdown
An [example application](https://github.com/facebookresearch/hydra/tree/master/plugins/hydra_joblib_launcher/example) using this launcher is provided in `plugins/hydra_joblib_launcher/example`. It changes the default launcher Hydra is using to this one.
```

Maybe I didn't get it, what is your suggestion?

Starting the app with `python my_app.py --multirun task=1,2,3,4,5` will launch five parallel executions.

Output of the example application:
```text
$ python example/my_app.py --multirun task=1,2,3,4,5
[2020-02-05 13:59:56,582][HYDRA] Sweep output dir : multirun/2020-02-05/13-59-56
[2020-02-05 13:59:56,584][HYDRA] Joblib.Parallel(n_jobs=-1,backend=None,prefer=processes,require=None,verbose=0,timeout=None,pre_dispatch=2*n_jobs,batch_size=auto,temp_folder=None,max_nbytes=1M,mmap_mode=r) is launching 5 jobs
[2020-02-05 13:59:56,584][HYDRA] Sweep output dir : multirun/2020-02-05/13-59-56
[2020-02-05 13:59:57,249][__main__][INFO] - Process ID 14336 executing task 2 ...
[2020-02-05 13:59:57,250][__main__][INFO] - Process ID 14333 executing task 1 ...
[2020-02-05 13:59:57,250][__main__][INFO] - Process ID 14334 executing task 3 ...
[2020-02-05 13:59:57,257][__main__][INFO] - Process ID 14335 executing task 4 ...
[2020-02-05 13:59:57,261][__main__][INFO] - Process ID 14337 executing task 5 ...

> ```

Reviewer: Probably better to exclude the timestamps from the output in the README, it adds too much noise.

Author: Done

# Hydra Joblib Launcher

> See [website](https://hydra.cc/docs/plugins/joblib_launcher) for more information

Reviewer: Until 1.0.0 is released, the url will be:

https://hydra.cc/docs/next/plugins/joblib_launcher
